src/gleam/time/timestamp.gleam

```diff
@@ -227,12 +265,388 @@ fn to_civil(minutes: Int) -> #(Int, Int, Int) {
+  // Done
+  use Nil <- result.try(accept_empty(bytes))
+
+  Ok(from_date_time(
+    year:,
+    month:,
+    day:,
+    hours:,
+    minutes:,
+    seconds:,
+    second_fraction_as_nanoseconds:,
+    offset_seconds:,
+  ))
+}
+
+fn parse_year(from bytes: BitArray) -> Result(#(Int, BitArray), Nil) {
+  parse_digits(from: bytes, count: 4)
+}
+
+fn parse_month(from bytes: BitArray) -> Result(#(Int, BitArray), Nil) {
+  use #(month, bytes) <- result.try(parse_digits(from: bytes, count: 2))
+  case 1 <= month && month <= 12 {
+    True -> Ok(#(month, bytes))
+    False -> Error(Nil)
+  }
+}
+
+fn parse_day(
+  from bytes: BitArray,
+  year year,
+  month month,
+) -> Result(#(Int, BitArray), Nil) {
+  use #(day, bytes) <- result.try(parse_digits(from: bytes, count: 2))
+
+  let is_leap_year = is_leap_year(year)
+
+  use max_day <- result.try(case month {
+    1 | 3 | 5 | 7 | 8 | 10 | 12 -> Ok(31)
+    4 | 6 | 9 | 11 -> Ok(30)
+    2 if is_leap_year -> Ok(29)
+    2 -> Ok(28)
+    _ -> Error(Nil)
+  })
+
+  case 1 <= day && day <= max_day {
+    True -> Ok(#(day, bytes))
+    False -> Error(Nil)
+  }
+}
+
+// Implementation from RFC 3339 Appendix C
+fn is_leap_year(year: Int) -> Bool {
+  year % 4 == 0 && { year % 100 != 0 || year % 400 == 0 }
+}
+
+fn parse_hours(from bytes: BitArray) -> Result(#(Int, BitArray), Nil) {
+  use #(hours, bytes) <- result.try(parse_digits(from: bytes, count: 2))
+  case 0 <= hours && hours <= 23 {
+    True -> Ok(#(hours, bytes))
+    False -> Error(Nil)
+  }
+}
+
+fn parse_minutes(from bytes: BitArray) -> Result(#(Int, BitArray), Nil) {
+  use #(minutes, bytes) <- result.try(parse_digits(from: bytes, count: 2))
+  case 0 <= minutes && minutes <= 59 {
+    True -> Ok(#(minutes, bytes))
+    False -> Error(Nil)
+  }
+}
+
+fn parse_seconds(from bytes: BitArray) -> Result(#(Int, BitArray), Nil) {
+  use #(seconds, bytes) <- result.try(parse_digits(from: bytes, count: 2))
+  // Max of 60 for leap seconds.  We don't bother to check if this leap second
+  // actually occurred in the past or not.
+  case 0 <= seconds && seconds <= 60 {
+    True -> Ok(#(seconds, bytes))
+    False -> Error(Nil)
+  }
+}
+
+// Truncates any part of the fraction that is beyond the nanosecond precision.
+fn parse_second_fraction_as_nanoseconds(from bytes: BitArray) {
+  case bytes {
+    <<".", byte, remaining_bytes:bytes>>
+      if byte_zero <= byte && byte <= byte_nine
+    -> {
+      do_parse_second_fraction_as_nanoseconds(
+        from: <<byte, remaining_bytes:bits>>,
+        acc: 0,
+        pow: nanoseconds_per_second,
+        k: 0,
+      )
+    }
+    // bytes starts with a ".", which should introduce a fraction, but it does
+    // not, and so it is an ill-formed input.
+    <<".", _:bytes>> -> Error(Nil)
+    // bytes does not start with a "." so there is no fraction.  Call this 0
+    // nanoseconds.
+    _ -> Ok(#(0, bytes))
+  }
+}
+
+fn do_parse_second_fraction_as_nanoseconds(from bytes, acc acc, pow pow, k k) {
+  case bytes {
```

Comment:
Collapse this nested case into one and favour > and < over the compare functions please